src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -82,6 +81,8 @@ def resolve(self, root_reqs, check_supported_wheels):
                 else:
                     constraints[name] = req.specifier
             else:
+                if req.is_direct and req.name:
+                    roots.add(canonicalize_name(req.name))
```

Comment:
I wonder if ther would be edge cases where an “unnamed” requirement messes up priority logic later. Would it be a good idea if we resolve all names eagerly here? We have to do it soon for resolvelib anyway.

Reply:
Honestly, I'm not sure. I'm really cautious about preparing too early, because it feels so easy to end up with another mess like canonical names<sup>1</sup>, where we keep going "let's resolve names now, because we don't know if it's a problem". That's why I like the new candidate approach of having a lazy property. Hopefully when we switch over to the project model, projects will do the same.

I'd like to understand *why* we need to resolve the name of an unnamed requirement here, if we do need to (we're talking about `pip install --upgrade /local/dir some other stuff` where we need to know that `/local/dir` is a root so that we know to upgrade it... uh, what?) Let's wait till something breaks, then we can write a test that clarifies *why* we need to do this, and *then* do it. At the moment, I'm not even sure what `pip install --upgrade /local/dir` should *mean*...

<sup>1</sup> I added that `canonicalize_name` call there as a result of a test failure, and it was another case of "just canonicalise, it's easier than trying to work out how come it's not already been done" 🙁